daemon.go

```diff
@@ -137,23 +139,51 @@ type providerOutput struct {
 	DataAvailableOverBitswap BitswapCheckOutput
 }
 
-// runCidCheck looks up the DHT for providers of a given CID and then checks their connectivity and Bitswap availability
-func (d *daemon) runCidCheck(ctx context.Context, cidStr string) (cidCheckOutput, error) {
-	cid, err := cid.Decode(cidStr)
+// runCidCheck finds providers of a given CID, using the DHT and IPNI
+// concurrently. A check of connectivity and Bitswap availability is performed
+// for each provider found.
+func (d *daemon) runCidCheck(ctx context.Context, cidKey cid.Cid, ipniURL string) (cidCheckOutput, error) {
+	crClient, err := client.New(ipniURL, client.WithStreamResultsRequired())
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to creat content router client: %w", err)
 	}
-
-	out := make([]providerOutput, 0, MaxProvidersCount)
+	routerClient := contentrouter.NewContentRoutingClient(crClient)
 
 	queryCtx, cancel := context.WithCancel(ctx)
 	defer cancel()
-	provsCh := d.dht.FindProvidersAsync(queryCtx, cid, MaxProvidersCount)
 
+	// Find providers with DHT and IPNI concurrently.
+	provsCh := d.dht.FindProvidersAsync(queryCtx, cidKey, MaxProvidersCount)
+	ipniProvsCh := routerClient.FindProvidersAsync(queryCtx, cidKey, MaxProvidersCount)
```

Comment:
Let's still pass `MaxProvidersCount` to each finder so that no one finds more than the max, but lets also cancel both and return as soon as we have collected a total of `MaxProvidersCount`

Reply:
SGTM